lms/services/lti_outcomes.py

```diff
@@ -0,0 +1,235 @@
+"""
+)
+
+# Template for a `readResult` XML message to send to the LMS.
+READ_RESULT_REQUEST_TEMPLATE = Template(
+    """
+<readResultRequest>
+  <resultRecord>
+    <sourcedGUID>
+      <sourcedId>{{ lis_result_sourcedid | e }}</sourcedId>
+    </sourcedGUID>
+  </resultRecord>
+</readResultRequest>
+"""
+)
+
+
+class LTIOutcomesRequestParams(NamedTuple):
+    """Common parameters used by all LTI Outcomes Management requests."""
+
+    consumer_key: str
+    """
+    OAuth 1.0 consumer key used to sign the request.
+    """
+
+    shared_secret: str
+    """
+    OAuth 1.0 shared secret used to sign the request.
+    """
+
+    lis_outcome_service_url: str
+    """
+    URL to submit requests to, provided by the LMS during an LTI launch.
+    """
+
+    lis_result_sourcedid: str
+    """
+    Opaque identifier for a particular submission.
+
+    This is provided by the LMS during an LTI launch and identifies the user
+    and assignment that the launch refers to.
+    """
+
+
+class LTIOutcomesClient:
+    """
+    Service for making requests to an LMS's Outcomes Management endpoint.
+
+    See https://www.imsglobal.org/specs/ltiomv1p0/specification.
+    """
+
+    def __init__(self, _context, request):
+        pass
+
+    def read_result(self, outcomes_request_params):  # noqa
+        """
+        Return the last-submitted score for a given submission.
+
+        :return: The last-submitted score or `None` if no score has been submitted.
+        """
+        body = READ_RESULT_REQUEST_TEMPLATE.render(
+            lis_result_sourcedid=outcomes_request_params.lis_result_sourcedid
+        )
+
+        result = _send_request(outcomes_request_params, body)
+
+        try:
+            score = find_element(
+                result, ["readResultResponse", "result", "resultScore", "textString"]
+            )
+            if score is None:
+                return None
+            return float(score.text)
+        except (ValueError, TypeError):
+            return None
+
+    def record_result(  # noqa
+        self,
+        outcomes_request_params,
+        score=None,
+        lti_launch_url=None,
+        submitted_at=None,
+    ):
+        """
+        Record a score or grading view launch URL for an assignment in the LMS.
+
+        :param score:
+            Float value between 0 and 1.0.
+            Defined as required by the LTI spec but is optional in Canvas if
+            an `lti_launch_url` is set.
+        :param lti_launch_url:
+            A URL where the student's work on this submission can be viewed.
+            This is only used in Canvas.
+        :param submitted_at:
+            A `datetime.datetime` that indicates when the submission was created.
+            This is only used in Canvas and is displayed in the SpeedGrader
+            as the submission date.
+            If the submission date matches an existing submission then the
+            existing submission is updated rather than creating a new submission.
```

Comment:
I think it is important to highlight here that we are dealing with an API that is, at least in principle, LMS independent, but that certain elements are Canvas-specific extensions to that base API which are definitely not supported in other LMSes.

If it turns out that we end up not using LTI to handle grading in other LMSes this will still be the case.